### PR TITLE
[numpy] Add `torch.xlogy`

### LINKED ISSUE #22656: Implement torch.xlogy
## 🚀 Feature
Implement `torch.xlogy` which computes `x * log(y)` if `x != 0` and 0 if `x == 0`.

## Motivation
Often one wants to compute `x * log(x)` or `x * log(y)` but define the result to be zero if x is 0 (instead of NaN). The current alternatives is to use torch.where to mask out NaN values or to add a small epsilon to x, like we do in binary cross entropy loss.

## Additional context

This is implemented in both SciPy and TensorFlow:

https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.xlogy.html
https://www.tensorflow.org/api_docs/python/tf/math/xlogy

cc @IssamLaradji

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #38349 
Fixes #22656

TODO:
* [x] Add docs
* [x] Add tests

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -2399,6 +2399,24 @@
   dispatch:
     DefaultBackend: logaddexp2
 
+- func: xlogy.Tensor(Tensor self, Tensor other) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:2402]: Add method, inplace, and out variants, too. 

Unfortunately this will double the number of signatures, since the inplace variant has a different name `xlogy_`.

File: test/test_autograd.py [hunk 2]
```diff
@@ -7277,6 +7277,54 @@ def test_atleast(self, device):
         self._test_atleast(device, torch.atleast_2d)
         self._test_atleast(device, torch.atleast_3d)
 
+    def test_xlogy(self, device):
```
Comment (mruberry) [inline test/test_autograd.py:7280]: grad and gradgrad checks are performed for every OpInfo in test_ops.py. Are those tests not sufficient for some reason? Is it because of xlogy's unique behavior at (0, 0)?

File: test/test_binary_ufuncs.py [hunk 3]
```diff
@@ -2495,6 +2499,55 @@ def _promo_helper(x, y):
                     with self.assertRaisesRegex(RuntimeError, "is not the desired type"):
                         torch.Tensor.float_power_(base.clone(), exp)
 
+    @skipIf(not TEST_SCIPY, "Scipy required for the test.")
+    @dtypes(*product(torch.testing.get_all_dtypes(include_complex=False, include_bfloat16=False),
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:2503]: xlogy is implemented for bfloat16, though

Testing a torch function that supports bfloat16 against a NumPy (or SciPy) reference is, unfortunately, a little tricky. It'd be nice if we had more helpers to support it. 

In this case, a special test for xlogy in bfloat16 could be added. It does not need to be as thorough as this test. For example, it does not need to test that out works properly, but this test (or another test) should verify that.

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -647,6 +647,16 @@
   self: grad / (1 + pow(2, other - self))
   other: grad / (1 + pow(2, self - other))
 
+- name: xlogy.Tensor(Tensor self, Tensor other) -> Tensor
+  self: grad * at::xlogy((self != 0), other)
```
Comment (mruberry) [inline tools/autograd/derivatives.yaml:651]: cc @albanD for this derivative

File: torch/_torch_docs.py [hunk 5]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
     {out}
 """.format(**common_args))
 
+add_docstr(torch.xlogy,
+           r"""
+xlogy(x, y) -> Tensor
+
+Compute x*log(y) so that the result is 0 if x = 0.
```
Comment (mruberry) [inline torch/_torch_docs.py:4279]: This is consistent with SciPy's documentation but misleading since xlogy(0, nan) is nan, not zero. 

Is that behavior an error, do you think? cc @rgommers. It would be interesting to look at this function's history to find out. It's a little odd that the following is true:

```
xlogy(0, nan) = 0 * log(nan) = 0 * nan = nan
xlogy(0, 0) = 0 * log(0) = 0 * -inf = 0
```

Since typically 0 * -inf is nan. So one case producing a nan is suppressed but another isn't.

File: torch/_torch_docs.py [hunk 6]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
...
+Similar to SciPy's `scipy.special.xlogy`.
+
+.. math::
+    \text{out}_i = x_i * \log{{y_i}}
```
Comment (mruberry) [inline torch/_torch_docs.py:4283]: I suggest making this 3 cases (given the current implementation): 

- y = Nan
- x = 0
- general case (this case)

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
...
+
+Args:
+    x (tensor): Multiplier
```
Comment (mruberry) [inline torch/_torch_docs.py:4286]: I suggest not bothering with giving these descriptions. This can just say:

```
self (Tensor)
other (Number or Tensor)
```

Note that both Number and Tensor are capitalized in this context.

File: torch/_torch_docs.py [hunk 8]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
...
+    y (float or tensor): Argument
+
+.. note::
```
Comment (mruberry) [inline torch/_torch_docs.py:4289]: See above discussion. If we keep this note move it so it's before the enumeration of arguments.

File: torch/_torch_docs.py [hunk 9]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
...
+    NaN in :attr:`y` is propagated even if value of :attr:`x` is 0.
+
+Example::
+
+    >>> x = torch.zeros(3,)
+    >>> y = torch.tensor([1, float('inf'), float('nan')])
```
Comment (mruberry) [inline torch/_torch_docs.py:4295]: Zero is another interesting  y value.

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
     {out}
 """.format(**common_args))
 
+add_docstr(torch.xlogy,
+           r"""
+xlogy(x, y) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:4277]: xlogy(self, other) to be consistent with other PyTorch functions (we're not looking at consistency with NumPy for named arguments)

File: torch/_torch_docs.py [hunk 11]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
...
+    tensor([0., 0., nan])
+    >>> x = torch.tensor([1, 2, 3])
+    >>> y = torch.tensor([3, 2, 1])
+    >>> torch.xlogy(x, y)
+    tensor([1.0986, 1.3863, 0.0000])
+    >>> torch.xlogy(x, 4)
+    tensor([1.3863, 2.7726, 4.1589])
+
+
+.. function:: xlogy(x, y) -> Tensor
```
Comment (mruberry) [inline torch/_torch_docs.py:4306]: I don't think this needs a second entry?

File: torch/_torch_docs.py [hunk 12]
```diff
@@ -4272,6 +4272,59 @@ def merge_dicts(*dicts):
     {out}
 """.format(**common_args))
 
+add_docstr(torch.xlogy,
```
Comment (mruberry) [inline torch/_torch_docs.py:4275]: Method variant will require modifying tensor_docs.py and .rst, too.

**Review by mruberry — CHANGES_REQUESTED** [review]
Hey @kshitij12345! Thanks for submitting this PR adding xlogy. The issue it fixes has been open for a long time. 

I made some comments/suggestions, especially about extending this to include method, inplace, and out= variants and using the OpInfo pattern.

Eventually we'll want to add a BinaryUfuncInfo class to the OpInfo architecture. But I think we could add xlogy as a regular OpInfo for now.

File: torch/_torch_docs.py (hunk 5 above)
Comment (rgommers) [reply torch/_torch_docs.py:4279]: Definitely not an error, returning `nan` is correct and was tested in the PR that created `xlogy`: https://github.com/scipy/scipy/pull/469.

The point of the function is that `0 * log(0)` should return `0`, this is a common special case in, e.g., statistics functions (and a valid mathematical limit value). `nan`s however simply propagate.

If the nan case bothers you, maybe just tweak this line of the docstring to say:
```
Compute ``x*log(y)`` so that ``0 * log(0)`` returns zero.
```

Comment (mruberry) [reply torch/_torch_docs.py:4279]: Thanks @rgommers for validating the current behavior, I agree we should tweak the docstring. Your proposal seems more precise.

File: test/test_autograd.py (hunk 2 above)
Comment (kshitij12345) [reply test/test_autograd.py:7169]: Yes. This is to particularly verify the behaviour at zero.

File: torch/_torch_docs.py (hunk 11 above)
Comment (kshitij12345) [reply torch/_torch_docs.py]: First entry covers two variants,
x -> tensor, y -> float, tensor

This entry covers,
x -> float, y -> tensor.

This is similar to `torch.pow`'s doc.
https://github.com/pytorch/pytorch/blob/a39398b9e5d528e4a6ca293f1703833932f0d9b2/torch/_torch_docs.py#L6403

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (kshitij12345) [reply aten/src/ATen/native/native_functions.yaml:2390]: Done.

File: test/test_binary_ufuncs.py (hunk 3 above)
Comment (kshitij12345) [reply test/test_binary_ufuncs.py:2503]: Added a new function for it.

File: torch/_torch_docs.py (hunk 5 above)
Comment (kshitij12345) [reply torch/_torch_docs.py]: Updated to 
`Compute ``x*log(y)`` so that ``0 * log(0)`` returns zero.`

File: torch/_torch_docs.py [hunk 13]
```diff
@@ -4338,6 +4338,80 @@ def merge_dicts(*dicts):
...
+           r"""
+xlogy(self, other, *, out=None) -> Tensor
+
+Compute ``self * log(other)`` so that ``0 * log(0)`` returns zero.
```
Comment (mruberry) [inline torch/_torch_docs.py:4345]: "Compute" -> "Computes"

"so that" -> "except treats \`\`0 * log(0)\`\` as zero."

File: torch/_torch_docs.py (hunk 8 above)
Comment (mruberry) [reply torch/_torch_docs.py:4358]: I think this note can be removed now that the cases are enumerated.

File: test/test_binary_ufuncs.py [hunk 14]
```diff
@@ -2495,6 +2499,105 @@ def _promo_helper(x, y):
...
+        _compare_helper(y, 3.14)
+        _compare_helper(z, 3.14)
+
+        # Special Values Tensor-Tensor
+        t = torch.tensor([0., 1., 2., float('inf'), -float('inf'), float('nan')], device=device)
+        zeros = torch.tensor(5, dtype=y_dtype, device=device)
+        _compare_helper(t, zeros)
+        _compare_helper(t, 0.)
+
+    def test_xlogy_exceptions(self, device):
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:2600]: I appreciate this test but it's testing TensorIterator more than xlogy, and I think we can remove it.

File: test/test_autograd.py (hunk 2 above)
Comment (mruberry) [reply test/test_autograd.py:7169]: This is really interesting, cc @albanD. I wonder if we can better support this use case in the future by extending OpInfos with metadata like "special_autograd_values."

The OpInfo test may also generate zeros, although it's not very likely. If it did, would those cause the regular autograd check to fail? I'm trying to understand what the OpInfo-based autograd tests can cover vs. what this extra test needs to validate.

File: torch/_torch_docs.py [hunk 15]
```diff
@@ -4338,6 +4338,80 @@ def merge_dicts(*dicts):
...
+    >>> x = torch.tensor([1, 2, 3])
+    >>> y = torch.tensor([3, 2, 1])
+    >>> torch.xlogy(x, y)
+    tensor([1.0986, 1.3863, 0.0000])
+    >>> torch.xlogy(x, 4)
+    tensor([1.3863, 2.7726, 4.1589])
+
+""".format(**common_args) + r"""
+
+.. function:: xlogy(self, other, *, out=None) -> Tensor
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/_torch_docs.py:4383]: First entry covers two variants,
x -> tensor, y -> float, tensor

Second entry covers,
x -> float, y -> tensor.

If we remove the second, then x -> float, y -> tensor won't be covered by doc. 

This is similar to `torch.pow`'s doc.
https://github.com/pytorch/pytorch/blob/a39398b9e5d528e4a6ca293f1703833932f0d9b2/torch/_torch_docs.py#L6403

[the remark this answers is not in the document]
Comment (mruberry) [reply torch/_torch_docs.py:4383]: Can they not be merged to describe that either x or y can be a Number or Tensor? It's true that this may mislead users into believing they can pass both inputs as Numbers, but that seems OK.

[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/_torch_docs.py:4383]: Cool. Will do that.

**Comment by kshitij12345** [comment]
@mruberry PTAL :)

File: aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu [hunk 16]
```diff
@@ -29,8 +30,23 @@ void mse_kernel_cuda(TensorIterator& iter) {
   });
 }
 
+void xlogy_kernel_cuda(TensorIterator& iter) {
+  AT_DISPATCH_FLOATING_TYPES_AND2(at::ScalarType::Half, at::ScalarType::BFloat16, iter.dtype(), "xlogy_cuda", [&]() {
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu:34]: Should this dispatch also be on common_dtype and not dtype()?

File: aten/src/ATen/native/native_functions.yaml [hunk 17]
```diff
@@ -2415,6 +2415,53 @@
...
+  use_c10_dispatcher: full
+  variants: function, method
+  dispatch:
+    DefaultBackend: xlogy
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:2422]: Because xlogy implements its own kernels, I think this should be `CPU, CUDA:` ?

File: torch/_torch_docs.py [hunk 18]
```diff
@@ -4338,6 +4338,48 @@ def merge_dicts(*dicts):
     {out}
 """.format(**common_args))
 
+add_docstr(torch.xlogy,
+           r"""
+xlogy(self, other, *, out=None) -> Tensor
+
+Computes ``self * log(other)`` except it treats ``0 * log(0)`` as ``0``.
```
Comment (mruberry) [inline torch/_torch_docs.py:4345]: I think we looked at this briefly before, but I don't think we can say "it treats..." because if `self` is zero then all values of `log`, including negative values, will return 0. And PyTorch treats the logarithm of negative numbers as nan, too:

```
t = torch.tensor((-2., -1, 0, 1, 2))
torch.log(t)
: tensor([   nan,    nan,   -inf, 0.0000, 0.6931])
```

The problem with the "it treats" clause here is that this implies it's the ONLY exception from the normal computation. 

What about saying, "Computes \`\`self * log(other)\`\` but with the following special cases:"

Then putting the mathematical cases below that line, then the sentence about this being similar to SciPy's xlogy after that?

File: torch/_torch_docs.py [hunk 19]
```diff
@@ -4338,6 +4338,48 @@ def merge_dicts(*dicts):
...
+        self_i * \log{(other_i)} & \text{otherwise}
+    \end{cases}
+
+""" + r"""
+
+Args:
+    self (Number or Tensor)
+    other (Number or Tensor)
+
+.. note:: Both attr:`self` and attr:`other` cannot be number.
```
Comment (mruberry) [inline torch/_torch_docs.py:4362]: "Both attr:`self` and attr:`other` cannot be number." -> "At least one of :attr:\`self\` or :attr:\`other\` must be a tensor."

File: torch/testing/_internal/common_methods_invocations.py [hunk 20]
```diff
@@ -859,6 +868,18 @@ def reference_sigmoid(x):
                                     dtypes=[torch.bfloat16]),),
                        assert_autodiffed=True,
                        promotes_integers_to_float=True),
+        OpInfo('xlogy',
+               dtypes=all_types_and(torch.bool),
+               dtypesIfCPU=all_types_and(torch.bool, torch.half, torch.bfloat16),
+               dtypesIfCUDA=all_types_and(torch.bool, torch.half, torch.bfloat16),
+               test_inplace_grad=True,
+               supports_tensor_out=True,
+               skips=(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:877]: This skip should be able to removed after a rebase

**Review by mruberry — COMMENTED** [review]
Hey @kshitij12345!

Overall this looks very good and thorough. Just a few small comments inline.

**Comment by mruberry** [comment]
The current coverage_test2 failure is unrelated to this PR, sorry about that.

Just let me know when this is ready for another review!

**Comment by kshitij12345** [comment]
@mruberry This is ready for review. Have addressed comments.

File: torch/_torch_docs.py [hunk 21]
```diff
@@ -4343,6 +4343,48 @@ def merge_dicts(*dicts):
...
+           r"""
+xlogy(self, other, *, out=None) -> Tensor
+
+Computes ``self * log(other)`` with the following cases.
+
+.. math::
+    \text{out}_{i} = \begin{cases}
+        \text{NaN} & \text{if } \text{other}_{i} = \text{NaN} \\
+        0 & \text{if } \text{self}_{i} = 0.0 \\
+        self_i * \log{(other_i)} & \text{otherwise}
```
Comment (mruberry) [inline torch/_torch_docs.py:4356]: `self_i` -> `\text{self}_i` and `other_i` -> `\text{other}_i` on this line

And one last change is that "self" should be renamed to "input" in this doc, sorry I forgot to mention that earlier. Just like how the parameters are named in add:

https://pytorch.org/docs/master/generated/torch.add.html?highlight=add#torch.add

File: torch/_torch_docs.py [hunk 22]
```diff
@@ -4343,6 +4343,48 @@ def merge_dicts(*dicts):
...
+    \end{cases}
+
+Similar to SciPy's `scipy.special.xlogy`.
+
+""" + r"""
+
+Args:
+    self (Number or Tensor)
```
Comment (mruberry) [inline torch/_torch_docs.py:4364]: "self" -> "input" here, too

File: torch/_torch_docs.py [hunk 23]
```diff
@@ -4343,6 +4343,48 @@ def merge_dicts(*dicts):
...
+    other (Number or Tensor)
+
+.. note:: At least one of :attr:`self` or :attr:`other` must be a tensor.
```
Comment (mruberry) [inline torch/_torch_docs.py:4367]: "self" -> "input" here as well

**Review by mruberry — APPROVED** [review]
Looks great, thanks @kshitij12345!

One last nit is the naming of the params (should be 'input', not 'self') and a bit of Latex formatting. Just ping me when that's done and I'll merge this.

**Comment by kshitij12345** [comment]
@mruberry PTAL :)

**Review by mruberry — APPROVED** [review]
Awesome! Thanks @kshitij12345!
